src/main/scala/io/iohk/ethereum/vm/ProgramState.scala

```diff
@@ -117,7 +117,7 @@ case class ProgramState[W <: WorldStateProxy[W, S], S <: Storage[S]](
   def toResult: ProgramResult[W, S] =
     ProgramResult[W, S](
       returnData,
-      if (error.isDefined) 0 else gas,
+      if (error.exists(_.useWholeGas)) 0 else gas,
```

Comment:
i think something similar should be done in https://github.com/input-output-hk/mantis/blob/phase/release2_0/src/main/scala/io/iohk/ethereum/ledger/BlockPreparator.scala#L124, as there we are calculating gas which we need to refund to sender.